src/util/circular_buffer.c

```diff
@@ -207,6 +207,49 @@ void *CircularBuffer_Read(CircularBuffer cb, void *item) {
   return read;
 }
 
+// Read all items from buffer.
+// dst must be large enough to hold all items in the buffer.
+// This function is not thread-safe.
+// This function pops the oldest item from the buffer.
+size_t CircularBuffer_ReadAll(CircularBuffer cb, void *dst, bool advance){
+  RedisModule_Assert(cb != NULL);
+
+  // make sure there's data to return
+  if (unlikely(CircularBuffer_Empty(cb))) {
+    return 0;
+  }
+
+  //calculate the offset_idx of the read pointer
+  size_t item_count = cb->item_count;
+  uint64_t write = cb->write;
+  uint idx = write / cb->item_size;
+  int read_idx = idx - item_count;
+  read_idx = (read_idx >= 0) ? read_idx : (cb->item_cap + read_idx);
+
+  size_t first_chunk = (read_idx + item_count <= cb->item_cap) ? item_count : (cb->item_cap - read_idx);
+  size_t second_chunk = item_count - first_chunk;
+  first_chunk *= cb->item_size;
+  second_chunk *= cb->item_size;
+
+  // copy item from buffer to output
+  void* read = cb->data + read_idx * cb->item_size;
+  if (dst != NULL) {
+    memcpy(dst, read, first_chunk);
```

Comment:
Note that this is a shallow copy - if we will have heap-allocated resources in our slow-log we will need either a deep copy or an ownership mechanism